dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/Java8IncompatibleReferenceCheck.java

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * A tool to find Java 8-incompatible references in the class files in a BOM. It checks the class
+ * files in the following artifacts:
+ *
+ * <ul>
+ *   <li>The artifacts in the {@code dependencyManagement} section of the BOM
+ *   <li>The artifacts in the dependency graph for each of the artifact above
+ *   <li>The latest version of the artifacts in the dependency graphs
+ * </ul>
+ *
+ * <p>If it finds references to classes in the core library, which are not present in Java 8, it
+ * exits with status code 1; otherwise it exits with status code 0.
+ *
+ * @see <a
+ *     href="https://github.com/GoogleCloudPlatform/cloud-opensource-java/wiki/Java-8-incompatible-references-of-java.nio.Buffer-classes-generated-by-Java-9--compilers"
+ *     >Java 8 incompatible references of java.nio.Buffer classes generated by Java 9 compilers</a>
+ */
+public class Java8IncompatibleReferenceCheck {
+
+  private static final Logger logger =
+      Logger.getLogger(Java8IncompatibleReferenceCheck.class.getName());
+
+  public static void main(String[] arguments)
+      throws MavenRepositoryException, IOException, URISyntaxException {
+
+    if (arguments.length != 1) {
+      System.err.println("Specify a path to the BOM file");
+      System.exit(1);
+    }
+
+    URI exclusionFileUri =
+        Java8IncompatibleReferenceCheck.class
+            .getClassLoader()
+            .getResource("java8-incompatible-reference-check-exclusion.xml")
+            .toURI();
+    Path exclusionFile = Paths.get(exclusionFileUri).toAbsolutePath();
+
+    String bomFileName = arguments[0];
+
+    Path bomFile = Paths.get(bomFileName);
+    Bom bom = Bom.readBom(bomFile);
+
+    ImmutableList<Artifact> managedDependencies = bom.getManagedDependencies();
+
+    int count = 1;
+
+    // The BOM member to problematic dependencies
+    ImmutableSetMultimap.Builder<Artifact, Artifact> problematicDependencies =
+        ImmutableSetMultimap.builder();
+    for (Artifact managedDependency : managedDependencies) {
+      logger.info(
+          "Checking "
+              + managedDependency
+              + " ("
+              + (count++)
+              + "/"
+              + managedDependencies.size()
+              + ")");
+
+      ClassPathBuilder classPathBuilder = new ClassPathBuilder();
+      ClassPathResult result = classPathBuilder.resolve(ImmutableList.of(managedDependency), false);
+
+      LinkageChecker linkageChecker =
+          LinkageChecker.create(result.getClassPath(), result.getClassPath(), exclusionFile);
+
+      ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
+
+      ImmutableSet<LinkageProblem> invalidJdkReferences =
+          linkageProblems.stream()
+              .filter(problem -> problem.getSymbol().getClassBinaryName().startsWith("java."))
+              .collect(toImmutableSet());
+
+      if (!invalidJdkReferences.isEmpty()) {
+        invalidJdkReferences.stream()
```

Comment:
This is what would contain the return type change?

Reply:
Yes. Renamed this to `invalidReferencesToJavaCoreLibrary`.